### PR TITLE
[13.0][ADD] account_move_reconcile_forbid_cancel

### PR DESCRIPTION
**Pull request by ernestotejeda**
Cc @Tecnativa  TT37518

This module extends the functionality of account invoicing to restrict 'reset to draft' or 'cancel' an invoice that has been reconciled.

### CODE REVIEW

**Comment by ernestotejeda** [comment]
ping @pedrobaeza @CarlosRoca13 @victoralmau

**Review by CarlosRoca13 — CHANGES_REQUESTED** [review]
Please change the commit message, you forgot the module name

**Comment by ernestotejeda** [comment]
thanks @CarlosRoca13 , done

**Comment by CarlosRoca13** [comment]
Now pre-commit is failing :fearful:

**Comment by ernestotejeda** [comment]
@CarlosRoca13 it's green now

File: account_move_reconcile_forbid_cancel/README.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/account-reconcile-13-0/account-reconcile-13-0-account_move_reconcile_forbid_cancel
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/98/13.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module extends the functionality of account invoicing to restrict
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/README.rst:28]:
```suggestion
This module restricts to cancel or reset to draft any invoice/journal entry that
has been reconciled (aka paid).
```

File: account_move_reconcile_forbid_cancel/README.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to *Invoicing > Customers > invoices*
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/README.rst:41]:
```suggestion
#. Go to *Invoicing > Customers > Invoices*
```

File: account_move_reconcile_forbid_cancel/README.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#. Edit an invoice that has a payment.
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/README.rst:42]:
```suggestion
#. Access an invoice that has a payment.
```

File: account_move_reconcile_forbid_cancel/README.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#. Click the 'Reset to Draft' button and you will get a validation error
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/README.rst:43]:
```suggestion
#. Click on the 'Reset to Draft' button and you will get an error
```

File: account_move_reconcile_forbid_cancel/README.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+   message to prevent you from resetting the invoice to draft state because
+   it already has an associated payment.
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/README.rst:45]:
```suggestion
   it has already an associated payment.
```

File: account_move_reconcile_forbid_cancel/models/account_move.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    def _get_receivable_payable_lines(self):
+        return self.line_ids.filtered(
+            lambda l: l.account_internal_type in ["receivable", "acc_type_pay"],
+        )
+
+    def button_draft(self):
+        rec_pay_lines = self._get_receivable_payable_lines()
+        if rec_pay_lines.matched_debit_ids or rec_pay_lines.matched_credit_ids:
+            raise ValidationError(_("You cannot reset to draft reconciled invoices."))
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/models/account_move.py]:
```suggestion
            raise ValidationError(_("You cannot reset to draft reconciled entries."))
```

File: account_move_reconcile_forbid_cancel/models/account_move.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        super().button_draft()
+
+    def button_cancel(self):
+        rec_pay_lines = self._get_receivable_payable_lines()
+        if rec_pay_lines.matched_debit_ids or rec_pay_lines.matched_credit_ids:
+            raise ValidationError(_("You cannot cancel reconciled invoices."))
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/models/account_move.py]:
```suggestion
            raise ValidationError(_("You cannot cancel reconciled entries."))
```

File: account_move_reconcile_forbid_cancel/tests/test_account_move_reconcile_forbid_cancel.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        cls.payment = payment_register.save()
+        cls.payment.post()
+
+    def test_reset_invoice_to_draft(self):
+        with self.assertRaises(ValidationError):
+            self.invoice.button_draft()
+
+    def test_cancel_invoice(self):
+        with self.assertRaises(ValidationError):
+            self.invoice.button_draft()
```
Comment (pedrobaeza) [inline account_move_reconcile_forbid_cancel/tests/test_account_move_reconcile_forbid_cancel.py]:
```suggestion
            self.invoice.button_cancel()
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by ernestotejeda** [comment]
@pedrobaeza changes done

**Comment by pedrobaeza** [comment]
You haven't changed translations.

**Comment by ernestotejeda** [comment]
@pedrobaeza , done

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**